service/frontend/api/producer_manager.go

```diff
@@ -0,0 +1,116 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+//go:generate mockgen -package $GOPACKAGE -source $GOFILE -destination producer_manager_mock.go -self_package github.com/uber/cadence/service/frontend/api
+
+package api
+
+import (
+	"fmt"
+
+	"github.com/uber/cadence/common/asyncworkflow/queue"
+	"github.com/uber/cadence/common/asyncworkflow/queue/provider"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/messaging"
+	"github.com/uber/cadence/common/metrics"
+)
+
+type (
+	ProducerManager interface {
+		GetProducerByDomain(domain string) (messaging.Producer, error)
+	}
+
+	producerManagerImpl struct {
+		domainCache   cache.DomainCache
+		provider      queue.Provider
+		logger        log.Logger
+		metricsClient metrics.Client
+
+		producerCache cache.Cache
+	}
+)
+
+func NewProducerManager(
+	domainCache cache.DomainCache,
+	provider queue.Provider,
+	logger log.Logger,
+	metricsClient metrics.Client,
+) ProducerManager {
+	return &producerManagerImpl{
+		domainCache:   domainCache,
+		provider:      provider,
+		logger:        logger,
+		metricsClient: metricsClient,
+		producerCache: cache.New(&cache.Options{
+			TTL:             0,
```

Comment:
having a 5-10 minutes TTL would be need to be able to support queue config changes at runtime and also to avoid keeping these in memory forever